### PR TITLE
feat(bigquery): add better timers around every API call

### PR DESCRIPTION
**Pull request by mayurinehate**
Tested with bigquery connector-tests for accuracy.

Primary changes -
1. Updated PerfTimer implementation to allow pausing timer
2. Renamed `BigQueryDataDictionary` to `BigQuerySchemaApi`, added timers for the calls in this class.
4. Refractored core audit log extraction methods into new class `BigQueryAuditLogApi`, added timers for the calls in this class.

Refractors:
1. Refractored connection config and related methods in `BigQueryConnectionConfig` class.
2. Refractored BigQueryDataDictionary to include bigquery client as instance method and moved related static methods to instance methods.
3. Refractored lineage code to be solely in `BigQueryLineageExtractor`.
4. Refractored and moved all queries/query templates to `bigquery_v2/queries.py`

Functional changes - 
1. Fix repetitive access to BQ exported audit logs in lineage case
2. Deduplicates insertId via QUALIFY in BQ exported audit logs in lineage case

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/utilities/perf_timer.py [hunk 1]
```diff
@@ -6,21 +6,53 @@
...
+
+        if self.end_time is not None:
+            self._past_active_time = self.elapsed_seconds()
+
         self.start_time = time.perf_counter()
         self.end_time = None
+        if self.paused:
+            self.paused = False
+
+    def pause_timer(self) -> "PerfTimer":
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/utilities/perf_timer.py:32]: Thoughts on renaming this to just `pause()`. Generally it's clear that the object being paused is a timer

File: metadata-ingestion/tests/unit/utilities/test_perf_timer.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+from datahub.utilities.perf_timer import PerfTimer
+
+
+def test_perf_timer_simple():
+    with PerfTimer() as timer:
+        time.sleep(1)
+        assert round(timer.elapsed_seconds()) == 1
+
+    assert round(timer.elapsed_seconds()) == 1
```
Comment (asikowitz) [inline metadata-ingestion/tests/unit/utilities/test_perf_timer.py:11]:
```suggestion
    assert pytest.approx(timer.elapsed_seconds()) == 1
```
Can do the same for others

**Review by asikowitz — COMMENTED** [review]
Had some trouble reviewing this because there's so much copied / moved code, ended up only skimming thru most of the lineage and usage changes. Looks very good to me, my main concerns around assertions in the perf timer. I think we should avoid erroring due to the perf time and instead note something went wrong in the report, if possible

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_audit_log_api.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        self,
+        bigquery_client: bigquery.Client,
+        bigquery_audit_metadata_query_template: Callable[
+            [
+                str,  # dataset: str
+                bool,  # use_date_sharded_tables: bool
+                Optional[int],  # limit: Optional[int] = None
+            ],
+            str,
+        ],
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_audit_log_api.py:48]: agree

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_audit_log_api.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+                if rate_limiter:
+                    with rate_limiter:
+                        for entry in query_job:
+                            with current_timer.pause_timer():
+                                yield entry
+                else:
+                    for entry in query_job:
+                        with current_timer.pause_timer():
+                            yield entry
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_audit_log_api.py:98]: refractored it slightly. I think it looks better now.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 5]
```diff
@@ -581,11 +561,11 @@ def _get_projects(self, conn: bigquery.Client) -> List[BigqueryProject]:
...
         else:
-            return list(self._get_project_list(conn))
+            return list(self._query_project_list())
 
-    def _get_project_list(self, conn: bigquery.Client) -> Iterable[BigqueryProject]:
+    def _query_project_list(self) -> Iterable[BigqueryProject]:
         try:
-            projects = BigQueryDataDictionary.get_projects(conn)
+            projects = self.bigquery_data_dictionary.get_projects()
         except Exception as e:
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:563]: not sure if it makes a big difference but I moved it as suggested.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_audit_log_api.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                filter_=filter,
+                page_size=log_page_size,
+                max_results=limit,
+            )
+
+            for i, entry in enumerate(list_entries):
+                if i == 0:
+                    logger.info(
+                        f"Starting log load from GCP Logging for {client.project}"
+                    )
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_audit_log_api.py:130]: removing this

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 7]
```diff
@@ -129,262 +126,52 @@ class BigqueryProject:
...
-    def get_query_result(conn: bigquery.Client, query: str) -> RowIterator:
+class BigQuerySchemaApi:
+    def __init__(
+        self, report: BigQuerySchemaApiPerfReport, client: bigquery.Client
+    ) -> None:
+        self.bq_client = client
+        self.report = report
+
+    def get_client(self) -> bigquery.Client:
+        assert self.bq_client is not None
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:137]: yep, removing this method altogether.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 8]
```diff
@@ -129,262 +126,52 @@ class BigqueryProject:
...
+        return self.bq_client
+
+    def get_query_result(self, query: str) -> RowIterator:
         logger.debug(f"Query : {query}")
-        resp = conn.query(query)
+        resp = self.get_client().query(query)
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:142]: yep

File: metadata-ingestion/src/datahub/utilities/perf_timer.py [hunk 9]
```diff
@@ -6,21 +6,53 @@
...
+
     """
 
-    start_time: Optional[float] = None
-    end_time: Optional[float] = None
+    def __init__(self) -> None:
+        self.start_time: Optional[float] = None
+        self.end_time: Optional[float] = None
+        self._past_active_time: float = 0
+        self.paused: Optional[bool] = None
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/utilities/perf_timer.py:16]: makes sense.

File: metadata-ingestion/src/datahub/utilities/perf_timer.py (hunk 1 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/utilities/perf_timer.py:32]: makes sense.

File: metadata-ingestion/src/datahub/utilities/perf_timer.py [hunk 10]
```diff
@@ -6,21 +6,53 @@
...
+        #    self.end_time is None
+        # ), "Can not start a finished timer. Did you accidentally re-use this timer ?"
+
+        if self.end_time is not None:
+            self._past_active_time = self.elapsed_seconds()
+
         self.start_time = time.perf_counter()
         self.end_time = None
+        if self.paused:
+            self.paused = False
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/utilities/perf_timer.py:30]: done

File: metadata-ingestion/src/datahub/utilities/perf_timer.py [hunk 11]
```diff
@@ -6,21 +6,53 @@
...
+
+    def pause_timer(self) -> "PerfTimer":
+        assert (
+            not self.paused and not self.end_time
+        ), "Can not pause a paused/stopped timer"
+        assert (
+            self.start_time is not None
+        ), "Can not pause a timer that hasn't started. Did you forget to start the timer ?"
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/utilities/perf_timer.py:38]: Right. Let me add warning logs in this case, and also add the error state indicator for reported  time in such case.

File: metadata-ingestion/tests/unit/utilities/test_perf_timer.py (hunk 2 above)
Comment (mayurinehate) [reply metadata-ingestion/tests/unit/utilities/test_perf_timer.py]: done

**Review by asikowitz — APPROVED** [review]

**Comment by asikowitz** [comment]
Looks like a legitimate failure on `FAILED tests/unit/test_bigquery_source.py::test_get_projects_list_failure` because of the project list change

**Comment by mayurinehate** [comment]
> Looks like a legitimate failure on `FAILED tests/unit/test_bigquery_source.py::test_get_projects_list_failure` because of the project list change

@asikowitz fixed the test.

**Pull request merged**